---
title: "Proteomics Data Analysis Report"
date: "`r format(Sys.time(), '%B %d, %Y')`"
always_allow_html: true
output:
  html_document: 
    toc: true
    toc_depth: 2
    toc_float: true
    theme: flatly
  bookdown::word_document2:
    toc: false
    
params:
  #progress info
  rendered_by_shiny: FALSE
  
  #Exploratory Analysis Report
  nSamples: NA
  rawdata: NA
  group: NA
  groupList: NA
  groupListConvert: NA
    
    #number of proteins detected
  nProteins: NA
  projName: NA
  speciesName: NA
  tissueT: NA
    
  sampleDistributionBar: NULL
  sampleDistributionDensity: NA
  original_heatmap: NA
  pcaParameter: NA
  summaryPCA: NA
  screePlot: NA
  pca3d: NA
  pca2d: NA
    
    
    #DE report
  result: NULL
  VolcanoPlotObject: NA
  DE_heatmapObject: NA
    
    
    
    #Enrichment Report
  en_result: NULL
  dotplot: NA
  enrichmentMap: NA
    
    
    #Normalization Report
  norSampleDistributionBar: NA
  norSampleDistributionDensity: NA
  batchList: NA
  batchListConvert: NA
  norData: NULL
  normed_heatmap: NA
  normedScreePlot: NA
  normedPCA3d: NA
  normedPCA2d: NA
  normed_pcaParameter: NA
    
    
  #Covariates Correction Report
  cov_group: NA
  aftercov: NULL
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, results='asis')
library(plotly)
library(knitr)
library(DT)
library(webshot)
if(is.null(webshot:::find_phantom())){webshot::install_phantomjs()}
```

If "Not available" is shown in the report, it means:\
You haven't generate the plot in JUMP-Suite. Only the plot was shown in JUMP-Suite at least one time, it can be output to the report.

------------------------------------------------------------------------

# Summary Page

Project Name: `r params$projName`

Date: `r Sys.time()`

## 1 Analyses Used in This Project:

```{r, analyses-run,echo=F, error=T}
if(!is.null(params$sampleDistributionBar)){
  cat("• Exploratory analysis\n")
}
if(nrow(params$norData)>0){
  cat("•	Normalization\n")
}
if(nrow(params$aftercov)>0){
  cat("•	Covariates Correction\n")
}
if(nrow(params$result)>0){
  cat("•	Differential gene expression analysis\n")
}
if(nrow(params$en_result)>0){
  cat("•	Enrichment Analysis\n")
}


```

------------------------------------------------------------------------

## 2 Input Data Summary:

-   Data Type: Mass spectrometry-based proteomics
-   Number of Samples: `r params$nSamples`
-   Number of Proteins: `r params$nProteins`
-   Species: `r params$speciesN`
-   Tissue Type: `r params$tissueT`

------------------------------------------------------------------------

## 3 Output Result Summary:

```{r output tables,echo=F, error=T}
Ntb <- 0
if(nrow(params$result)>0) Ntb = Ntb+1
if(nrow(params$en_result)>0) Ntb = Ntb+1
if(nrow(params$norData)>0) Ntb = Ntb+1
if(nrow(params$aftercov)>0) Ntb = Ntb+1
```

Number of tables: `r Ntb`

```{r number of figures,echo=F, error=T}
Nplt <- 0
if(!is.null(params$VolcanoPlotObject)) Nplt = Nplt+1
if(!is.null(params$DE_heatmapObject)) Nplt = Nplt+1
if(!is.null(params$norSampleDistributionBar)) Nplt = Nplt+1
if(!is.null(params$original_heatmap)) Nplt = Nplt+1
if(!is.null(params$screePlot)) Nplt = Nplt+1
if(!is.null(params$pca3d)) Nplt = Nplt+1
if(!is.null(params$pca2d)) Nplt = Nplt+1
if(!is.null(params$dotplot)) Nplt = Nplt+1
if(!is.null(params$normed_heatmap)) Nplt = Nplt+1
if(!is.null(params$normedScreePlot)) Nplt = Nplt+1
if(!is.null(params$normedPCA3d)) Nplt = Nplt+1
if(!is.null(params$normedPCA2d)) Nplt = Nplt+1
```

Number of plots: `r Nplt`

------------------------------------------------------------------------

# Exploratory Analysis Report

## 1 Input Summary

```{r input summary}
#kable(head(params$rawdata, 5), caption = "First 5 rows of original dataset.")
datatable(
  params$rawdata,
  caption = "Raw Data.",
  extensions = c("Scroller","RowReorder","Buttons"),
  options = list(
    pageLength = 5,
    autoWidth = TRUE,
    dom = "Bfrtip",
    scrollY = 500,
    scroller = TRUE,
    scrollX = TRUE,
    buttons = list(list(
      extend = 'csv',
      filename = paste0("raw_data_", Sys.Date())
    ))
  )
)
```

**Size of original dataset:**

`r nrow(params$rawdata)` rows\
`r ncol(params$rawdata)` columns

**Number of group:** `r length(params$groupList)`

**Group information:**

```{r groupList}
for(i in 1:length(params$groupList)){
  group_name <- names(params$groupList)[i]
  cat(paste0(group_name, ": ",
             paste0(params$groupList[[i]], collapse = ", "), 
             "  \n")
      )
}
```

**List of group infomation is shown as below:**

```{r groupListConvert, results='markup'}
params$groupListConvert
```

------------------------------------------------------------------------

# Exploratory Result

## 1 Density Bar Plot and Density Plot

```{r density plot}
if(!is.null(params$sampleDistributionBar)){
  params$sampleDistributionBar
}else{
  cat("Sample distribution bar plot not available")
}

if(!is.null(params$sampleDistributionDensity)){
  params$sampleDistributionDensity
}else{
  cat("Sample distribution density plot not available")
}
```

## 2 Pincipal Component Analysis(PCA)

The PCA parameters is shown as below:

```{r pca para}
if(!is.null(params$pcaParameter)){
# Get parameters from TCC-GUI
pcaTransform <- params$pcaParameter$pcaTransform
pcaTopGene <- params$pcaParameter$pcaTopGene
pcaCenter <- params$pcaParameter$pcaCenter
pcaScale <- params$pcaParameter$pcaScale

pcaPara <- rbind(
  c("pcaTopGene", "Number of gene", pcaTopGene),
  c("pcaTransform", "Log(x+1) transform", pcaTransform),
  c("pcaCenter", "Center", pcaCenter),
  c("pcaScale", "Scale", pcaScale)
  )
  colnames(pcaPara) <-
  c("Parameter", "Explanation", "Value")
  kable(pcaPara, caption = "Paramters setting of PCA.")
} else {
  cat("Not available")
}
```

<!-- ```{r pca computation} -->

<!-- if(length(params$pcaParameter) > 0){ -->

<!-- cat("Using top N gene to perform PCA\n") -->

<!-- cat("```r\n") -->

<!-- cat("# Obtain dataset from TCC Object\n") -->

<!-- cat("data <- tcc$count\n") -->

<!-- # PCA processing -->

<!-- if (pcaTransform == TRUE) { -->

<!--   cat("\n# Log transform dataset\n") -->

<!--   cat("data <- log1p(data)\n") -->

<!-- } -->

<!-- cat("\n# Perform PCA, Extract genes with none-zero variance\n") -->

<!-- cat("data <- data[apply(data, 1, var) != 0, ]\n") -->

<!-- if(!is.na(pcaTopGene) & pcaTopGene < nrow(params$tccObject$count)){ -->

<!--   cat("data <- t(data[order(apply(data, 1, var), decreasing = TRUE)[1:", pcaTopGene, "], ])\n") -->

<!-- } -->

<!-- cat(paste0("data.pca <- prcomp(data[, apply(data, 2, var) != 0], center = ",  -->

<!--            pcaCenter, -->

<!--            ", scale. = ",  -->

<!--            pcaScale, ")\n")) -->

<!-- cat("```\n") -->

<!-- } else { -->

<!--   cat("") -->

<!-- } -->

<!-- ``` -->

### Summary Table of PCA

```{r summary table of PCA}
if(!is.null(params$summaryPCA)){
  params$summaryPCA
} else {
  "Not available"
}
```

### Scree plot

```{r}
if(!is.null(params$screePlot)){
  params$screePlot
} else {
  "Not available"
}
```

### 3D

```{r}
if(!is.null(params$pca3d)){
  params$pca3d
} else {
  "Not available"
}
```

### 2D

```{r}
if(!is.null(params$pca2d)){
  params$pca2d
} else {
  "Not available"
}
```

### Heatmap

```{r}
if(!is.null(params$pca2d)){
  params$original_heatmap
} else {
  "Not available"
}
```

------------------------------------------------------------------------

# Normalization

If normalization if performed, results will be shown. Group using for normalization is shown as below.

Number of group: `r length(params$batchList)`

```{r norm parameters}
for(i in 1:length(params$batchList)){
  group_name <- names(params$batchList)[i]
  cat(paste0(group_name, ": ",
             paste0(params$batchList[[i]], collapse = ", "), 
             "  \n")
      )
}

```

**Size of normalized dataset**:\
`r nrow(params$norData)` rows\
`r ncol(params$norData)` columns

```{r show normalized data}
#kable(head(params$norData, 5), caption = "First 10 row of normalized dataset.")
datatable(
  params$norData,
  caption = "Normalized Data.",
  extensions = c("Scroller","RowReorder","Buttons"),
  options = list(
    pageLength = 5,
    autoWidth = TRUE,
    dom = "Bfrtip",
    scrollY = 500,
    scroller = TRUE,
    scrollX = TRUE,
    buttons = list(list(
      extend = 'csv',
      filename = paste0("normalized_data_", Sys.Date())
    ))
  )
)
```

**List of batch infomation is shown as below**:

```{r batchListConvert, results='markup'}
params$batchListConvert
```

------------------------------------------------------------------------

## 1 Sample Distribution

```{r}
if(!is.null(params$norSampleDistributionDensity)){
  params$norSampleDistributionDensity
} else {
  "Not available"
}
```

## 2 Principal Components Analysis

The PCA parameters is shown as below:

```{r normed pca para}
if(!is.null(params$normed_pcaParameter)){
# Get parameters from JUMP
pcaTransform <- params$normed_pcaParameter$pcaTransform
pcaTopGene <- params$normed_pcaParameter$pcaTopGene
pcaCenter <- params$normed_pcaParameter$pcaCenter
pcaScale <- params$normed_pcaParameter$pcaScale

pcaPara <- rbind(
  c("pcaTopGene", "Number of gene", pcaTopGene),
  c("pcaTransform", "Log(x+1) transform", pcaTransform),
  c("pcaCenter", "Center", pcaCenter),
  c("pcaScale", "Scale", pcaScale)
  )
  colnames(pcaPara) <-
  c("Parameter", "Explanation", "Value")
  kable(pcaPara, caption = "Paramters setting of PCA.")
} else {
  cat("Not available")
}
```

### Scree plot

```{r}
if(!is.null(params$normedScreePlot)){
  params$normedScreePlot
} else {
  "Not available"
}
```

### 3D

```{r}
if(!is.null(params$normedPCA3d)){
  params$normedPCA3d
} else {
  "Not available"
}
```

### 2D

```{r}
if(!is.null(params$normedPCA2d)){
  params$normedPCA2d
} else {
  "Not available"
}
```

## Heatmap

```{r}
if(!is.null(params$normed_heatmap)){
  params$normed_heatmap
} else {
  "Not available"
}
```

------------------------------------------------------------------------

# Covariates Correction

The groups you selected for correction: `r params$cov_group`

## Result Table

```{r}
if(!is.null(params$corrected_data)){
  #kable(head(params$corrected_data, 10), caption = "First 10 rows of covariates corrected data")
  datatable(
  params$corrected_data,
  caption = "Covariates Corrected Data.",
  extensions = c("Scroller","RowReorder","Buttons"),
  options = list(
    pageLength = 5,
    autoWidth = TRUE,
    dom = "Bfrtip",
    scrollY = 500,
    scroller = TRUE,
    scrollX = TRUE,
    buttons = list(list(
      extend = 'csv',
      filename = paste0("cov_corrected_data_", Sys.Date())
    ))
  )
)
  
}else{
  cat("Not Available")
}

```

------------------------------------------------------------------------

# Differential Expression Analysis

## 1 Result Table

Here is a preview of your differential analysis results.

```{r}
if (params$rendered_by_shiny)
  shiny::setProgress(0.5)  # set progress to 50%

if(nrow(params$result)>0){
  #kable(head(params$result, 10), caption = "First 10 rows of differential expression result table")
  datatable(
  params$result,
  caption = "Differential Analysis Result.",
  extensions = c("Scroller","RowReorder","Buttons"),
  options = list(
    pageLength = 5,
    autoWidth = TRUE,
    dom = "Bfrtip",
    scrollY = 500,
    scroller = TRUE,
    scrollX = TRUE,
    buttons = list(list(
      extend = 'csv',
      filename = paste0("Differential_Analysis_Result_", Sys.Date())
    ))
  )
)
  
}else{
  cat("Not Available")
}

```

## 2 Volcano Plot

```{r}
if(!is.null(params$VolcanoPlotObject)){
  params$VolcanoPlotObject
} else {
  "Not available"
}
```

## 3 Heatmap

```{r}
if(!is.null(params$DE_heatmapObject)){
  params$DE_heatmapObject
} else {
  "Not available"
}
```

------------------------------------------------------------------------

# Enrichment Analysis

## 1 Enrichment Analysis Summary

Here is a preview of your enrichment results.

```{r}
if(is.null(params$en_result)){
  #kable(head(params$en_result@result, 10), caption = "First 10 row of enrichment result table")
  datatable(
  params$en_result@result,
  caption = "Enrichemtn Analysis Result.",
  extensions = c("Scroller","RowReorder","Buttons"),
  options = list(
    pageLength = 5,
    autoWidth = TRUE,
    dom = "Bfrtip",
    scrollY = 500,
    scroller = TRUE,
    scrollX = TRUE,
    buttons = list(list(
      extend = 'csv',
      filename = paste0("Enrichment_Analysis_Result_", Sys.Date())
    ))
  )
)
}else{
  "Not available"
}

```

## 2 Dot Plot

```{r}
if(!is.null(params$dotplot)){
  params$dotplot
} else {
  "Not available"
}
```

## 3 Enrichment Map

```{r}
if(!is.null(params$enrichmentMap)){
  params$enrichmentMap
} else {
  "Not available"
}

if (params$rendered_by_shiny)
  shiny::setProgress(1)  # set progress to 100%
```
